### PR TITLE
Recalculate lerp if we got infinity. Eliminates some overflows.

### LINKED ISSUE #1917: <cmath>: `lerp(1e+308, 5e+307, 4.0)` spuriously overflows
**Command-line test case**
```
C:\Users\He\source\test>type test.cpp
```
```cpp
#include <cmath>
#include <format>
#include <iostream>
using namespace std;

int main() {
    cout << format("expected: {}\n", -1e+308);
    cout << format("  actual: {}\n", lerp(1e+308, 5e+307, 4.0));
    return 0;
}
```
```
C:\Users\He\source\test>cl /EHsc /std:c++latest test.cpp
用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.29.30031 版
版权所有(C) Microsoft Corporation。保留所有权利。

/std:c++latest 作为最新的 C++
working 草稿中的语言功能预览提供。我们希望你提供有关 bug 和改进建议的反馈。
但是，请注意，这些功能按原样提供，没有支持，并且会随着工作草稿的变化
而更改或移除。有关详细信息，请参阅
https://go.microsoft.com/fwlink/?linkid=2045807。

test.cpp
Microsoft (R) Incremental Linker Version 14.29.30031.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:test.exe
test.obj

C:\Users\He\source\test>.\test.exe
expected: -1e+308
  actual: -inf
```

**STL version**
https://github.com/microsoft/STL/commit/65eb5077778a11cc4abac88b5fd5d0a25a2dcd19

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #1917

### CODE REVIEW

**Comment by fsb4000** [comment]
@statementreply 
```
It breaks monotonicity.
```
Thank you for finding some issues. I fixed some but I have no idea how to restore monotonicity for lerp :(

**Comment by statementreply** [comment]
> I fixed some but I have no idea how to restore monotonicity for lerp :(

Never mind. It can be tricky to implement floating point functionalities when there are requirements on precision, handling of extreme values, and/or preservation of math properties.

I'm going to open a PR (maybe on the weekend) with an alternative approach to fix the false overflow.

File: stl/inc/cmath [hunk 1]
```diff
@@ -1330,6 +1340,36 @@ _NODISCARD constexpr _Ty _Common_lerp(const _Ty _ArgA, const _Ty _ArgB, const _T
...
+            return fetestexcept(FE_OVERFLOW) != 0;
+#else // ^^^ defined(_M_FP_STRICT) / !defined(_M_FP_STRICT) vvv
+            return _Is_inf(val);
+#endif // _M_FP_STRICT
+        };
+
+        auto _Clear_FE_OVERFLOW_flag = []() constexpr {
+#ifdef _M_FP_STRICT
+            if (!_STD is_constant_evaluated()) {
+                feclearexcept(FE_OVERFLOW);
```
Comment (statementreply) [inline stl/inc/cmath:1357]: This clears `FE_OVERFLOW` if it was set before calling `lerp`. Unlike `errno`, IEEE 754 exception flags are sticky, and library functions should not clear any flags that were set before calling them.

There are ways to hide unwanted internal floating-point exceptions (see e.g. [`feholdexcept`](https://en.cppreference.com/w/cpp/numeric/fenv/feholdexcept)), but I would recommend against manipulating floating-point exceptions if you are not familiar with them or numerical algorithms in general.

**Comment by abolz** [comment]
FWIW, this _could_ be fixed with a fused multiply-add
```diff
--const auto _Candidate = _ArgA + _ArgT * (_ArgB - _ArgA);
++const auto _Candidate = std::fma(_ArgT, _ArgB - _ArgA, _ArgA);
```
But `fma` is not constexpr.

File: stl/inc/cmath [hunk 2]
```diff
@@ -1330,6 +1374,36 @@ _NODISCARD constexpr _Ty _Common_lerp(const _Ty _ArgA, const _Ty _ArgB, const _T
...
+            }
+#endif // _M_FP_STRICT
+        };
+
+        auto _Signbit = [](_Ty val) constexpr {
+            // std::signbit is not constexpr so cannot use :(
+            return val < 0;
```
Comment (cbezault) [inline stl/inc/cmath:1398]: Is this a legitimate way to check the sign bit? I haven't looked at the rest of the code to determine if it matters.

Comment (cbezault) [reply stl/inc/cmath:1398]: [Maybe  just copy our testing utils?](https://github.com/microsoft/STL/blob/62137922ab168f8e23ec1a95c946821e24bde230/tests/std/tests/GH_000935_complex_numerical_accuracy/floating_point_utils.hpp#L44)

Comment (fsb4000) [reply stl/inc/cmath]: It seems only for -0.0 the functions return different results.
```
signbit(-0.0) = 1
-0.0 < 0 = 0
```
And it doesn't matter.
But sure, we can copy the existing code.

File: stl/inc/cmath [hunk 3]
```diff
@@ -23,9 +23,20 @@
...
+#include <cfenv>
+#endif // _M_FP_STRICT
+#include <type_traits>
 #include <xutility>
 #endif // _HAS_CXX20
 
+#if _HAS_CXX20
+#ifdef _M_FP_STRICT
+// TRANSITION, VSO-923474 -- should be #pragma STDC FENV_ACCESS ON
+#pragma fenv_access(on)
```
Comment (statementreply) [inline stl/inc/cmath:36]: This would affect user code and unrelated STL code.

File: stl/inc/cmath [hunk 4]
```diff
@@ -1304,6 +1315,40 @@ template <class _Ty>
 _NODISCARD constexpr _Ty _Common_lerp(const _Ty _ArgA, const _Ty _ArgB, const _Ty _ArgT) noexcept {
     // on a line intersecting {(0.0, _ArgA), (1.0, _ArgB)}, return the Y value for X == _ArgT
 
+#ifdef _M_FP_STRICT
+    class _ExceptGuard {
```
Comment (statementreply) [inline stl/inc/cmath:1319]: It's also allowed to change rounding mode under `/fp:precise` mode with `#pragma fenv_access(on)`.

Floating-point mode can be changed by [`#pragma float_control`](https://docs.microsoft.com/en-us/cpp/preprocessor/float-control?view=msvc-160), in which case `_M_FP_***` macros doesn't reflect the actual mode in effect.

File: stl/inc/cmath [hunk 5]
```diff
@@ -1304,6 +1315,40 @@ template <class _Ty>
...
+        }
+
+        _ExceptGuard(const _ExceptGuard&) = delete;
+        _ExceptGuard& operator=(const _ExceptGuard&) = delete;
+
+        constexpr ~_ExceptGuard() {
+            if (!_STD is_constant_evaluated()) {
+                [[maybe_unused]] const int _UpdateEnv = feupdateenv(&_Env);
+            }
+        }
```
Comment (statementreply) [inline stl/inc/cmath:1334]: `feholdexcept` and `feupdateenv` are quite costly on x86 CPUs. The use case here is one of the scenarios IEEE 754 exceptions were originally designed for, but I think we'd better avoid them for short computations like `lerp` because of performance.

As an example, floating-point `midpoint` checks whether `_Val1 + _Val2` could possibly overflow before actually adding them.
https://github.com/microsoft/STL/blob/18c12ab01896e73e95a69ceba9fbd7250304f895/stl/inc/numeric#L632-L648

I did a simple x64 benchmark on AMD Ryzen 3700X. `a`, `b`, and `t` are random `double`s within [0, 1]:
Algorithm | Time/ns
:---: | :---:
Naïve `a * (1-t) + b * t` | &#x2007;&#x2007;0.493
`std::lerp` before PR (https://github.com/microsoft/STL/commit/18c12ab01896e73e95a69ceba9fbd7250304f895) | &#x2007;&#x2007;7.35&#x2007;
`std::lerp` after PR (https://github.com/microsoft/STL/pull/1918/commits/eb6cc26fc97a057c7540a91abe6c0b3af82087ce) | 135&#x2008;&#x2007;&#x2007;&#x2007;

**Comment by fsb4000** [comment]
@statementreply

Well, at least we can reuse new test code :(

What do you think about fma? It seems it doesn't overflow but we can get different results for runtime/compiletime invocation of lerp...

https://gcc.godbolt.org/z/T564z58qo

File: stl/inc/cmath [hunk 6]
```diff
@@ -1318,7 +1319,14 @@ _NODISCARD constexpr _Ty _Common_lerp(const _Ty _ArgA, const _Ty _ArgB, const _T
...
 
         // exact at _ArgT == 0, monotonic except near _ArgT == 1, bounded, determinate, and consistent:
-        const auto _Candidate = _ArgA + _ArgT * (_ArgB - _ArgA);
+        const auto _Candidate = [=]() constexpr {
+            if (_STD is_constant_evaluated()) {
+                return _ArgA + _ArgT * (_ArgB - _ArgA);
+            }
+            return _STD fma(_ArgT, _ArgB - _ArgA, _ArgA);
+        }
+        ();
```
Comment (fsb4000) [inline stl/inc/cmath:1328]: clang-format doesn't understand immediately invoked lambda :(

File: stl/inc/cmath (hunk 2 above)
Comment (cbezault) [reply stl/inc/cmath]: Sorry for the bother 😀. Like I said, I didn't really check if that would cause a problem.

**Comment by MattStephanson** [comment]
The condition we have to handle here is limited to  |`t*(b-a)`| &le; `2*MEOW_MAX`, right? Otherwise we'd have a true, not spurious, overflow. If that's the case, maybe we can simply scale the calculation by a fixed factor.
```cpp
auto _Smaller     = _ArgT;
auto _Larger      = _ArgB - _ArgA;
auto _Abs_smaller = _Float_abs(_Smaller);
auto _Abs_larger  = _Float_abs(_Larger);
if (_Abs_larger < _Abs_smaller) {
    _STD swap(_Smaller, _Larger);
    _STD swap(_Abs_smaller, _Abs_larger);
}

if (_Abs_larger > _STD sqrt(numeric_limits<_Ty>::max()) && _Abs_smaller > 1) {
    return 2 * (_Ty{0.5} * _ArgA + _Smaller * (_Ty{0.5} * _Larger));
} else {
    return _ArgA + _Smaller * _Larger;
}
```
`_Larger` is too large to be subnormal, so scaling by 0.5 is exact, and the product `_Smaller * _Larger` is large enough that if `_ArgA` is subnormal, it will be too small to contribute anyway. So I think the two expressions should be equivalent, apart from internal overflow, so there won't be a loss of monotonicity when switching from one to the other.

**Comment by statementreply** [comment]
> What do you think about fma? It seems it doesn't overflow but we can get different results for runtime/compiletime invocation of lerp...

- Accuracy

  `fma(t, b - a, a)` is (on average) more accurate than `a + t * (b - a)`, and doesn't suffer from `t * (b - a)` overflowing as long as the final result doesn't overflow. `(1 - t) * a + t * b` in the other branch could also be `fma(t, b, fma(-t, a, a))`.

- Performance

  Recent CPUs (AMD since ~2012, Intel since ~2014, and all ARM that Windows arm64 supports) have hardware fma instructions. Older CPUs don't support hardware fma , so `std::fma` needs to be implemented in software.

  On the current version of MSVC, if not compiled with `/arch:AVX2` or alike, `std::fma` isn't inlined and generates a call into UCRT (https://godbolt.org/z/zaY6veW9j). It then checks whether the CPU supports fma. If so, it calls the CPU instruction directly (however the function call overhead is non-trivial). If not, UCRT uses another algorithm to compute the result, which is unfortunately neither accurate nor fast as of ucrtbase.dll 10.0.19041.789.

- Different results between compile time and runtime

  There are already situations where <cmath> results could differ.

  `<cmath>` functions (`exp`, `sin`, etc.) use fma powered implementation on CPUs with hardware fma, and a different implementation without fma on CPUs without hardware fma. Code using these math functions could generate different results on different CPUs.

  GCC performs constant folding (not limited to `constexpr`) on math functions. The compile time results could be different from the results from the runtime math library. (https://godbolt.org/z/r49WEax6E)

  Also, the compiler is allowed to compile `x * y + z` into `fma(x, y, z)` (https://godbolt.org/z/zaY6veW9j), which may also lead to different results between compile time and runtime.

**Comment by statementreply** [comment]
> The condition we have to handle here is limited to |`t*(b-a)`| ≤ `2*MEOW_MAX`, right? Otherwise we'd have a true, not spurious, overflow. If that's the case, maybe we can simply scale the calculation by a fixed factor.
> 
> ```c++
> auto _Smaller     = _ArgT;
> auto _Larger      = _ArgB - _ArgA;
> auto _Abs_smaller = _Float_abs(_Smaller);
> auto _Abs_larger  = _Float_abs(_Larger);
> if (_Abs_larger < _Abs_smaller) {
>     _STD swap(_Smaller, _Larger);
>     _STD swap(_Abs_smaller, _Abs_larger);
> }
> 
> if (_Abs_larger > _STD sqrt(numeric_limits<_Ty>::max()) && _Abs_smaller > 1) {
>     return 2 * (_Ty{0.5} * _ArgA + _Smaller * (_Ty{0.5} * _Larger));
> } else {
>     return _ArgA + _Smaller * _Larger;
> }
> ```
>
> `_Larger` is too large to be subnormal, so scaling by 0.5 is exact, and the product `_Smaller * _Larger` is large enough that if `_ArgA` is subnormal, it will be too small to contribute anyway. So I think the two expressions should be equivalent, apart from internal overflow, so there won't be a loss of monotonicity when switching from one to the other.

Note that we have |_b_ - _a_| <= `MEOW_MAX` in this branch, so it is limited to |_t_| > 1. I think we could do the following to save a few branches (untested, needs benchmark):
```cpp
T lerp(T a, T b, T t) {
    // [...]
    if (abs(t) < 1) {
        T candidate = a + t * (b - a);
        // [...]
    }

    if (t == 1) {
        return b;
    }

    T half_prod = slope * (T{0.5} * t);
    if (abs(half_prod) <= numeric_limits<T>::max() * T{0.5}) {
        return intercept + slope * t;
    } else {
        return T{2} * (T{0.5} * intercept + half_prod);
    }
}
```

File: stl/inc/cmath (hunk 6 above)
Comment (BillyONeal) [reply stl/inc/cmath:1328]: I would prefer to see this given a named function because naming it tends to reduce duplication between translation units.

File: tests/std/tests/P0811R3_midpoint_lerp/test.cpp [hunk 7]
```diff
@@ -1023,6 +1023,48 @@ bool test_lerp() {
     return true;
 }
 
+void test_GH_1917() {
```
[the remark this answers is not in the document]
Comment (fsb4000) [reply tests/std/tests/P0811R3_midpoint_lerp/test.cpp:1026]: I think the test will not benefit from `floating_point_model_matrix.lst`

**Comment by fsb4000** [comment]
I think you are correct about speed of `std::fma` I didn't test speed but it produces correct results for `lerp(1e+308, 5e+307, 4.0)` so I think it is better then previous direct multiplication and addition `_ArgA + _ArgT * (_ArgB - _ArgA);`

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/cmath [hunk 8]
```diff
@@ -1335,6 +1337,40 @@ _NODISCARD auto hypot(const _Ty1 _Dx, const _Ty2 _Dy, const _Ty3 _Dz) {
...
+        auto _Smaller     = _ArgT;
+        auto _Larger      = _ArgB - _ArgA;
+        auto _Abs_smaller = _Float_abs(_Smaller);
+        auto _Abs_larger  = _Float_abs(_Larger);
+        if (_Abs_larger < _Abs_smaller) {
+            _STD swap(_Smaller, _Larger);
+            _STD swap(_Abs_smaller, _Abs_larger);
+        }
+
+        if (_Abs_smaller > 1 && _Abs_larger > _Floating_max<remove_cv_t<_Ty>> / _Abs_larger) {
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/cmath:1364]: I see that @MattStephanson in https://github.com/microsoft/STL/pull/1918#issuecomment-841042608 originally suggested
```cpp
if (_Abs_larger > _STD sqrt(numeric_limits<_Ty>::max()) && _Abs_smaller > 1) {
```
so this is a rephrasing that avoids `sqrt`, but it still seems to me that
```cpp
if (_Abs_smaller > 1 && _Abs_larger > _Floating_max<remove_cv_t<_Ty>> / _Abs_smaller) {
```
would be better to test here. Again, I could be wrong.

[the remark this answers is not in the document]
Comment (fsb4000) [reply stl/inc/cmath]: Yes, you are correct. That will activate this codepath more often than necessary but it is probably OK, because the two path should give the same result when no overflow.
I think it was just a slightly more rough estimate and not typo.
But I think you are right and we can replace it with `_Abs_smaller`.
But just in case, I will also call @MattStephanson , if you have time could you take a look?

[the remark this answers is not in the document]
Comment (MattStephanson) [reply stl/inc/cmath]: The test suggested by @StephanTLavavej is more precise and appears to be at least mostly correct (it might be possible that `_Abs_larger` is slightly larger than the exact quotient, but the the quotient gets rounded up to be equal to `_Abs_larger` and the fast path is mistakenly taken.) Nonetheless, I have concern about the test. (1) This is the `constexpr` path, so performance is less critical; (2) the `sqrt(max)` could be precalculated; and (3) I doubt that always doing 1 division to sometimes save 3 multiples is a good tradeoff.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/cmath (hunk 8 above)
[the remark this answers is not in the document]
Comment (fsb4000) [reply stl/inc/cmath]: I like the idea because now we can remove the additional include files.

File: stl/inc/cmath [hunk 9]
```diff
@@ -1335,6 +1335,31 @@ _NODISCARD auto hypot(const _Ty1 _Dx, const _Ty2 _Dy, const _Ty3 _Dz) {
...
+        if (_Abs_larger < _Abs_smaller) {
+            _STD swap(_Smaller, _Larger);
+            _STD swap(_Abs_smaller, _Abs_larger);
+        }
+
+        // _Larger is too large to be subnormal, so scaling by 0.5 is exact, and the product _Smaller * _Larger is large
+        // enough that if _ArgA is subnormal, it will be too small to contribute anyway and this way can sometimes avoid
+        // overflow problems
+        if (_Abs_smaller > 1) {
+            return 2 * (_Ty{0.5} * _ArgA + _Smaller * (_Ty{0.5} * _Larger));
```
Comment (StephanTLavavej) [inline stl/inc/cmath:1354]: This is a great explanation, thanks! One last suggestion - this is describing something that is only true when the `if` branch is taken. Therefore, the comment should be within the `if` branch.

(If the comment were phrased conditionally, then it would be fine to place before the condition.)

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/cmath (hunk 8 above)
[the remark this answers is not in the document]
Comment (MattStephanson) [reply stl/inc/cmath]: > Can/should we simplify the condition to just _Abs_smaller > 1?

FWIW, I think this is safe. You're correct in pointing out that there are two separate issues, performance and correctness, and the correctness of the `0.5`-scaled path depends on not underflowing. I think that ensuring that both multiplicands are at least 1 covers that.

**Comment by StephanTLavavej** [comment]
I'm speculatively mirroring this to the MSVC-internal repo. Further changes can be pushed, but please notify me.

File: stl/inc/cmath [hunk 10]
```diff
@@ -1335,6 +1335,31 @@ _NODISCARD auto hypot(const _Ty1 _Dx, const _Ty2 _Dy, const _Ty3 _Dz) {
...
+            // sometimes avoid overflow problems.
+            return 2 * (_Ty{0.5} * _ArgA + _Smaller * (_Ty{0.5} * _Larger));
+        } else {
+            return _ArgA + _Smaller * _Larger;
+        }
+    }
+
+    return _STD fma(_ArgT, _ArgB - _ArgA, _ArgA);
+}
+
```
Comment (CaseyCarter) [inline stl/inc/cmath:1362]: We need to find that Twitter thread mocking our huge implementation of `lerp` to let them know we've added another 25 lines. (No change requested.)

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for infinitely improving this implementation! :infinity: :rocket: :joy_cat:
